### PR TITLE
Implement a file system TemplateProvider and ModuleTypeProvider

### PR DESCRIPTION
**Pull request by adimova**
implements issue #1569
replace PR https://github.com/eclipse/smarthome/pull/1648
Signed-off-by: Ana Dimova a.dimova@prosyst.bg

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/build.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+output.. = target/classes/
+bin.includes = META-INF/,\
+               .,\
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/build.properties]: What ebout ESH-INF and OSGI-INF?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+ * automation objects.
+ *
+ * @author Ana Dimova - Initial Contribution
+ *
+ */
+@SuppressWarnings("rawtypes")
+public abstract class AbstractFileProvider<E> extends AbstractWatchService implements Provider<E> {
+
+    protected Logger logger = LoggerFactory.getLogger(this.getClass());
+    protected String[] WatchingDirs;
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Member variables should start with a lower case.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    protected String WatchingDir;
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Member variables should start with a lower case.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+    @Override
+    public Collection<E> getAll() {
+        synchronized (providedObjectsHolder) {
+            return providedObjectsHolder.values();
+        }
+    }
+
+    @Override
+    public void removeProviderChangeListener(ProviderChangeListener<E> listener) {
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Can we move this function below the add function (just for easier readability).

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+    protected void removeParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.remove(parserType);
+    }
+
+    protected void importResources(File file) {
+        if (file.exists()) {
+            File[] files = file.listFiles();
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: I don't know, but need we to check if file is a directory (or perhaps a symlink to a directory) first?

Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Ah, okay, I assume if null is returned you check that implicitly.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+                }
+            }
+        }
+    }
+
+    protected void remove(File path) {
+        try {
+            URL url = path.toURI().toURL();
+            synchronized (providerPortfolio) {
+                removeElements(providerPortfolio.remove(url));
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Does this remove files and files in the given directory?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+                    updateProvidedObjectsHolder(url, providedObjects);
+                }
+            } catch (ParsingException e) {
+                logger.debug(e.getMessage(), e);
+            } catch (IOException e) {
+                logger.debug(e.getMessage(), e);
+            } finally {
+                try {
+                    if (inputStreamReader != null) {
+                        inputStreamReader.close();
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Can this close throw an exception? If it is possible then is is not closed.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractMultipleFileProvider.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        implements ProviderChangeListener<E>, Provider<E> {
+
+    protected static final String ROOTS = "roots";
+
+    protected Map<String, P> providers;
+    protected String[] roots = new String[] { "automation" };
+    protected Map<Parser<E>, Map<String, String>> parsers = new HashMap<Parser<E>, Map<String, String>>();
+    protected List<ProviderChangeListener<E>> listeners = new ArrayList<ProviderChangeListener<E>>();
+
+    public void deactivate() {
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractMultipleFileProvider.java]: Isn't a deactivate method defined in "AbstractFileProvider", too?
Don't you need to use the override method and call `super.deactivate()`?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                        provider.addProviderChangeListener(this);
+                        provider.activate(parsers);
+                        providers.put(this.roots[i], provider);
+                    }
+                }
+            }
+        }
+    }
+
+    public void activate(Map<String, Object> config) {
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java]: You extend `AbstractMultipleFileProvider` that extends `AbstractFileProvider`. Don't we need to call that activate method?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+ * @author Ana Dimova - Initial Contribution
+ *
+ */
+public class ModuleTypeFileProvider extends AbstractFileProvider<ModuleType> implements ModuleTypeProvider {
+
+    private final String moduleTypesRoot = "moduletypes";
+
+    public ModuleTypeFileProvider(String root) {
+        WatchingDir = root + File.separator + moduleTypesRoot;
+        super.activate();
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java]: If you call activate in a constructor, who ever calls deactivate?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    public void activate(Map<String, Object> config) {
+        modified(config);
+        providers = new HashMap<String, TemplateFileProvider>(this.roots.length);
+        for (int i = 0; i < roots.length; i++) {
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java]: Isn't that loop logic already be done in the modified method and needs only be done if it is not already handled?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+ * @author Ana Dimova - Initial Contribution
+ *
+ */
+public class TemplateFileProvider extends AbstractFileProvider<RuleTemplate> implements TemplateProvider {
+
+    private final String templatesRoot = "templates";
+
+    public TemplateFileProvider(String root) {
+        WatchingDir = root + File.separator + templatesRoot;
+        super.activate();
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java]: See question above

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/build.properties (hunk 1 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/build.properties]: Ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 2 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 3 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 4 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 6 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Thanks for catching this - it is working only for files now, I assumed that when a directory is deleted, I'll receive events for all files in it. I'll fix it

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 7 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractMultipleFileProvider.java (hunk 8 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractMultipleFileProvider.java]: AbstractMultipleFileProvider does not extend AbstractFileProvider, it's a holder of multiple AbstractFileProvider instances, so it must call deactivate() on each of them. I'll fix it

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java (hunk 9 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java]: AbstractMultipleFileProvider does not extend AbstractFileProvider; I'll move the call to AbstractWatchService.activate() from the constructors of ModuleTypeFileProvider and TemplateFileProvider, to the activate method of AbstractFileProvider.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java (hunk 10 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java]: See the deactivate comment above

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java (hunk 11 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java]: I was trying to optimize the activation, but if you don't like it, I can simplify the code and remove the duplications

**Comment by maggu2810** [comment]
@danchom Can you review this PR, too?

**Comment by danchom** [comment]
Yes of course.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
+    
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true"
+    name="org.eclipse.smarthome.automation.internal.provider.file.ModuleTypeFileProvider">
+    <implementation class="org.eclipse.smarthome.automation.internal.provider.file.MultipleModuleTypeFileProvider" />
+    <reference bind="addParser" cardinality="0..n" interface="org.eclipse.smarthome.automation.parser.Parser" name="Parser" policy="static" unbind="removeParser" target="(parser.type=parser.module.type)"/>
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml]: I'm not sure if the cardinality="0..n" and policy="static"  are the right combination.
Usually "0..n" cardinality is combined with dynamic policy. What do you think?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+ * automation objects.
+ *
+ * @author Ana Dimova - Initial Contribution
+ *
+ */
+@SuppressWarnings("rawtypes")
+public abstract class AbstractFileProvider<E> extends AbstractWatchService implements Provider<E> {
+
+    protected Logger logger = LoggerFactory.getLogger(this.getClass());
+    protected String[] watchingDirs;
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: This field watchingDirs is not used anywhere.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+                }
+                notifyListeners(mtRemoved);
+            }
+        }
+    }
+
+    private String getParserType(URL url) {
+        String fileName = url.getPath();
+        int fileExtesionStartIndex = fileName.lastIndexOf(".") + 1;
+        if (fileExtesionStartIndex == -1) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: fileExtesionStartIndex will never be -1 even when there is not an file extension.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            }
+        }
+        for (int i = 0; i < this.roots.length; i++) {
+            if (!providers.containsKey(this.roots[i])) {
+                initializeProvider(this.roots[i]);
+            }
+        }
+    }
+
+    public void activate(Map<String, Object> config) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java]: I think activate method should be move in the parent class. There can be define abstract modified(Map<String, Object> config) method too.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            }
+        }
+        for (int i = 0; i < this.roots.length; i++) {
+            if (!providers.containsKey(this.roots[i])) {
+                initializeProvider(this.roots[i]);
+            }
+        }
+    }
+
+    public void activate(Map<String, Object> config) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java]: Move the activate method  in the parent class.

**Comment by adimova** [comment]
The PR #2196 fixes the comments

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml (hunk 13 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml]: OK

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 14 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java (hunk 16 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleModuleTypeFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java (hunk 17 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/MultipleTemplateFileProvider.java]: ok

**Comment by danchom** [comment]
LGTM except that if I remove the entire root dirs of moduleTypes/templates the corresponding moduleTypes/templates which are put into these dirs are not remove from the RuleEngine.
But I'm not sure if this is a bug.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T extends ModuleType> Collection<T> getModuleTypes(Locale locale) {
+        Collection<ModuleType> values = providedObjectsHolder.values();
+        if (values.isEmpty()) {
+            return Collections.<T>emptyList();
+        }
+        Collection<ModuleType> toReturn = new LinkedList<ModuleType>();
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java]: You probably could directly do
```
Collection<ModuleType> toReturn = new LinkedList<ModuleType>(values);
```

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T extends Template> Collection<T> getTemplates(Locale locale) {
+        Collection<RuleTemplate> values = providedObjectsHolder.values();
+        if (values.isEmpty()) {
+            return Collections.<T>emptyList();
+        }
+        Collection<RuleTemplate> toReturn = new LinkedList<RuleTemplate>();
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java]: You probably could directly do
```
Collection<RuleTemplate> toReturn = new LinkedList<RuleTemplate>(values);
```

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true"
+    name="org.eclipse.smarthome.automation.internal.provider.file.ModuleTypeFileProvider">
+    <implementation class="org.eclipse.smarthome.automation.internal.provider.file.ModuleTypeFileProviderWatcher" />
+    <reference bind="addParser" cardinality="1..n" interface="org.eclipse.smarthome.automation.parser.Parser" name="Parser" policy="dynamic" unbind="removeParser" target="(parser.type=parser.module.type)"/>
+    <service>
+        <provide interface="org.eclipse.smarthome.automation.type.ModuleTypeProvider" />
+    </service>
+    <property name="service.config.description.uri" type="String" value="system:automationresources:moduletypes"/>
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml]: as mentioned above, better remove these properties

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+    
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true"
+    name="org.eclipse.smarthome.automation.internal.provider.file.TemplateFileProvider">
+    <implementation class="org.eclipse.smarthome.automation.internal.provider.file.TemplateFileProviderWatcher" />
+    <reference bind="addParser" cardinality="1..n" interface="org.eclipse.smarthome.automation.parser.Parser" name="Parser" policy="dynamic" unbind="removeParser" target="(parser.type=parser.template)"/>
+    <service>
+        <provide interface="org.eclipse.smarthome.automation.template.TemplateProvider" />
+    </service>
+    <property name="service.config.description.uri" type="String" value="system:automationresources:templates"/>
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml]: as mentioned above, better remove these properties

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 22]
```diff
@@ -0,0 +1,308 @@
+/**
+ * Copyright (c) 1997, 2015 by ProSyst Software GmbH and others.
```
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+    protected void importFile(String parserType, URL url) {
+        Parser<E> parser = parsers.get(parserType);
+        if (parser != null) {
+            InputStream is = null;
+            InputStreamReader inputStreamReader = null;
+            try {
+                is = url.openStream();
+                BufferedInputStream bis = new BufferedInputStream(is);
+                inputStreamReader = new InputStreamReader(bis);
+                Set<E> providedObjects = parser.parse(inputStreamReader);
```
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+            } catch (ParsingException e) {
+                logger.debug(e.getMessage(), e);
+            } catch (IOException e) {
+                logger.debug(e.getMessage(), e);
+            } finally {
+                try {
+                    if (inputStreamReader != null) {
+                        inputStreamReader.close();
+                    }
+                    if (is != null) {
```
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: it is in a finally block

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+    protected abstract void initializeWatchService(String watchingDir);
+
+    protected abstract void deactivateWatchService(String watchingDir);
+
+    private String getParserType(URL url) {
+        String fileName = url.getPath();
+        if (fileName.lastIndexOf(".") == -1) {
+            return Parser.FORMAT_JSON;
+        }
+        String fileExtesion = fileName.substring(fileName.lastIndexOf(".") + 1);
```
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AutomationWatchService.java [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+
+    private String watchingDir;
+    private AbstractFileProvider provider;
+
+    public AutomationWatchService(AbstractFileProvider provider, String watchingDir) {
+        this.provider = provider;
+        this.watchingDir = watchingDir;
+        File file = new File(watchingDir);
+        if (!file.exists()) {
+            file.mkdirs();
```
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AutomationWatchService.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java (hunk 18 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 24 above)
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Yes in "a", but not in the one of the try-catch block the code is in.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 25 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AutomationWatchService.java (hunk 26 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AutomationWatchService.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java (hunk 19 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml (hunk 20 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml (hunk 21 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml]: ok

File: bundles/automation/pom.xml [hunk 27]
```diff
@@ -20,6 +20,7 @@
     <module>org.eclipse.smarthome.automation.commands</module>
     <module>org.eclipse.smarthome.automation.parser.gson</module>
     <module>org.eclipse.smarthome.automation.providers</module>
+	<module>org.eclipse.smarthome.automation.provider.file</module>
```
Comment (kaikreuzer) [inline bundles/automation/pom.xml]: could you please replace the tab by spaces (or alternatively all spaces by tabs)?

File: docs/documentation/development/rules/ModuleTypes_schema.json [hunk 28]
```diff
@@ -1,965 +1,1029 @@
 {
-	"$schema":"http://json-schema.org/draft-04/schema#",
```
Comment (kaikreuzer) [inline docs/documentation/development/rules/ModuleTypes_schema.json]: I was just wondering - did we agree on tabs vs spaces in JSON files already - in general I would tend to say we should do the same as for XML (as the files are packaged within the bundles), i.e. we should go for tabs. But I see that the current formatter uses spaces and thus has reformatted all your files.

**Comment by maggu2810** [comment]
We didn't set a format, yet.
The only comments I remember is that the currently (that time) used three could be changed to four spaces or tabs: https://github.com/eclipse/smarthome/pull/1639

**Comment by kaikreuzer** [comment]
@adimova Could you update the indentation issues? We'd then good to merge.

**Comment by adimova** [comment]
Done

**Comment by kaikreuzer** [comment]
Hm, now all schemas are shown as fully changed. @adimova Could you please revert this and only add you change in here? If we want to re-format the json files altogether, we should do that in a separate commit/PR and probably only after we have set auto-formatting settings in the IDE setup.

**Comment by adimova** [comment]
Now you should be able to see the differences, but the files are not correctly formatted

**Comment by kaikreuzer** [comment]
It still shows many unnecessary changes (e.g. https://github.com/eclipse/smarthome/pull/2070/files#diff-8265eb2de84c90f78aa5762a537114fdR434 ff). Can't you simply revert those?

**Comment by adimova** [comment]
Sorry i can't see unnecessary changes

**Comment by kaikreuzer** [comment]
Sorry, my mistake; I just saw hundreds of new lines in the schema and thought they were due to formatting changes (and I didn't expect huge changes in the context of this PR to the schemas - it seems these parts were simply missing in the past?).

PR is ok for me now, but as it is more than 1000 lines, I cannot merge it. But as you work in the same company as @danchom, he is allowed to merge it without requiring a CQ :-)

**Comment by adimova** [comment]
thanks and yes the new lines were missing in the past

**Pull request merged**